### ISSUE TITLE
Cannot import torch on latest nightly

### ISSUE THREAD

**Issue by rafi-cohen** [pytorch/pytorch#56882]
## 🐛 Bug

Importing torch on the latest nightly results in the following error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "path/to/envs/nightly/lib/python3.8/site-packages/torch/__init__.py", line 196, in <module>
    _load_global_deps()
  File "path/to/envs/nightly/lib/python3.8/site-packages/torch/__init__.py", line 149, in _load_global_deps
    ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
  File "path/to/envs/nightly/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libmkl_intel_lp64.so.1: cannot open shared object file: No such file or directory
```

## To Reproduce

1. `conda create -n nightly python=3.8 pytorch torchvision torchaudio cudatoolkit=11.1 -c pytorch-nightly -c conda-forge` (you might want to change the environment name)
1. `conda activate nightly` (or whatever name you chose)
1. `python -c "import torch"`

## Expected behavior

`torch` imports successfuly

## Environment

```
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.8 (64-bit runtime)
Is CUDA available: N/A
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti
GPU 4: GeForce RTX 2080 Ti
GPU 5: GeForce RTX 2080 Ti
GPU 6: GeForce RTX 2080 Ti
GPU 7: GeForce RTX 2080 Ti

Nvidia driver version: 460.32.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.9.0.dev20210424
[pip3] torchaudio==0.9.0a0+71318ea
[pip3] torchvision==0.10.0.dev20210424
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.1.1               h6406543_8    conda-forge
[conda] mkl                       2020.4             h726a3e6_304    conda-forge
[conda] mkl-service               2.3.0            py38h1e0a361_2    conda-forge
[conda] mkl_fft                   1.3.0            py38h5c078b8_1    conda-forge
[conda] mkl_random                1.2.0            py38hc5bc63f_1    conda-forge
[conda] numpy                     1.19.2           py38h54aff64_0
[conda] numpy-base                1.19.2           py38hfa32c7d_0
[conda] pytorch                   1.9.0.dev20210424 py3.8_cuda11.1_cudnn8.0.5_0    pytorch-nightly
[conda] torchaudio                0.9.0.dev20210424            py38    pytorch-nightly
[conda] torchvision               0.10.0.dev20210424      py38_cu111    pytorch-nightly
```

## Additional context

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @seemethere @malfet @walterddr

**Comment by NicolasHug** [comment]
A similar issue on MacOS:

```bash
➜  ~ conda create -n nightly python=3.8 pytorch -c pytorch-nightly && conda activate nightly && python -c "import torch"
Collecting package metadata (current_repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.9.2
  latest version: 4.10.1

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /Users/nicolashug/opt/miniconda3/envs/nightly

  added / updated specs:
    - python=3.8
    - pytorch


The following NEW packages will be INSTALLED:

  blas               pkgs/main/osx-64::blas-1.0-mkl
  ca-certificates    pkgs/main/osx-64::ca-certificates-2021.4.13-hecd8cb5_1
  certifi            pkgs/main/osx-64::certifi-2020.12.5-py38hecd8cb5_0
  intel-openmp       pkgs/main/osx-64::intel-openmp-2021.2.0-hecd8cb5_564
  libcxx             pkgs/main/osx-64::libcxx-10.0.0-1
  libffi             pkgs/main/osx-64::libffi-3.3-hb1e8313_2
  libuv              pkgs/main/osx-64::libuv-1.40.0-haf1e3a3_0
  mkl                pkgs/main/osx-64::mkl-2021.2.0-hecd8cb5_269
  mkl-service        pkgs/main/osx-64::mkl-service-2.3.0-py38h9ed2024_1
  mkl_fft            pkgs/main/osx-64::mkl_fft-1.3.0-py38h4a7008c_2
  mkl_random         pkgs/main/osx-64::mkl_random-1.2.1-py38hb2f4e1b_2
  ncurses            pkgs/main/osx-64::ncurses-6.2-h0a44026_1
  ninja              pkgs/main/osx-64::ninja-1.10.2-hf7b0b51_1
  numpy              pkgs/main/osx-64::numpy-1.20.1-py38hd6e1bb9_0
  numpy-base         pkgs/main/osx-64::numpy-base-1.20.1-py38h585ceec_0
  openssl            pkgs/main/osx-64::openssl-1.1.1k-h9ed2024_0
  pip                pkgs/main/osx-64::pip-21.0.1-py38hecd8cb5_0
  python             pkgs/main/osx-64::python-3.8.8-h88f2d9e_5
  pytorch            pytorch-nightly/osx-64::pytorch-1.9.0.dev20210425.arm64-py3.8_0
  readline           pkgs/main/osx-64::readline-8.1-h9ed2024_0
  setuptools         pkgs/main/osx-64::setuptools-52.0.0-py38hecd8cb5_0
  six                pkgs/main/osx-64::six-1.15.0-py38hecd8cb5_0
  sqlite             pkgs/main/osx-64::sqlite-3.35.4-hce871da_0
  tk                 pkgs/main/osx-64::tk-8.6.10-hb0a8c7a_0
  typing_extensions  pkgs/main/noarch::typing_extensions-3.7.4.3-pyha847dfd_0
  wheel              pkgs/main/noarch::wheel-0.36.2-pyhd3eb1b0_0
  xz                 pkgs/main/osx-64::xz-5.2.5-h1de35cc_0
  zlib               pkgs/main/osx-64::zlib-1.2.11-h1de35cc_3


Proceed ([y]/n)?

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate nightly
#
# To deactivate an active environment, use
#
#     $ conda deactivate

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/nicolashug/opt/miniconda3/envs/nightly/lib/python3.8/site-packages/torch/__init__.py", line 196, in <module>
    _load_global_deps()
  File "/Users/nicolashug/opt/miniconda3/envs/nightly/lib/python3.8/site-packages/torch/__init__.py", line 149, in _load_global_deps
    ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
  File "/Users/nicolashug/opt/miniconda3/envs/nightly/lib/python3.8/ctypes/__init__.py", line 381, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: dlopen(/Users/nicolashug/opt/miniconda3/envs/nightly/lib/python3.8/site-packages/torch/lib/libtorch_global_deps.dylib, 10): no suitable image found.  Did find:
	/Users/nicolashug/opt/miniconda3/envs/nightly/lib/python3.8/site-packages/torch/lib/libtorch_global_deps.dylib: mach-o, but wrong architecture
	/Users/nicolashug/opt/miniconda3/envs/nightly/lib/python3.8/site-packages/torch/lib/libtorch_global_deps.dylib: mach-o, but wrong architecture
```

These issues seem to be the cause of the CI failures in torchvision (https://github.com/pytorch/vision/issues/3716, https://github.com/pytorch/vision/issues/3696) and torchaudio

**Comment by gulsmyigit** [comment]
I have a similar issue when I run the following colab notebook. 
[https://colab.research.google.com/github/patil-suraj/exploring-T5/blob/master/T5_on_TPU.ipynb#scrollTo=knpacarPX2AN](url)

**Comment by skyline75489** [comment]
I can easily repro this. This is the result of `readelf -d`:

```text
❯ readelf -d libtorch_cpu.so

Dynamic section at offset 0x89bf6c0 contains 39 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libgomp.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [librt.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libmkl_intel_lp64.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libmkl_gnu_thread.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libmkl_core.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libcudart.so.11.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc10.so]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
 0x000000000000000e (SONAME)             Library soname: [libtorch_cpu.so]
```

The actual binary comes with `conda install mkl` has names like `libmkl_intel_lp64.so`, without the `.1` suffix.

**Comment by seemethere** [comment]
Closing https://github.com/pytorch/pytorch/issues/57209 in favor of this one

**Comment by seemethere** [comment]
Regarding the original failure this seems to be related to conda-forge, when I did a `s/conda-forge/nvidia/` to pull `cudatoolkit=11.1` from `-c nvidia` it appears to work:
<details>
<summary>Click for logs</summary>

```
/tmp/tmp.A1nJh47HsJ
base ❯ conda create -y -p $(pwd) -c pytorch-nightly -c nvidia python=3.8 pytorch cudatoolkit=11.1 && conda activate $(pwd)
Collecting package metadata (current_repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.10.0
  latest version: 4.10.1

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /tmp/tmp.A1nJh47HsJ

  added / updated specs:
    - cudatoolkit=11.1
    - python=3.8
    - pytorch


The following NEW packages will be INSTALLED:

  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main
  blas               pkgs/main/linux-64::blas-1.0-mkl
  ca-certificates    pkgs/main/linux-64::ca-certificates-2021.4.13-h06a4308_1
  certifi            pkgs/main/linux-64::certifi-2020.12.5-py38h06a4308_0
  cudatoolkit        nvidia/linux-64::cudatoolkit-11.1.74-h6bb024c_0
  intel-openmp       pkgs/main/linux-64::intel-openmp-2021.2.0-h06a4308_610
  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.33.1-h53a641e_7
  libffi             pkgs/main/linux-64::libffi-3.3-he6710b0_2
  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0
  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0
  libuv              pkgs/main/linux-64::libuv-1.40.0-h7b6447c_0
  mkl                pkgs/main/linux-64::mkl-2021.2.0-h06a4308_296
  mkl-service        pkgs/main/linux-64::mkl-service-2.3.0-py38h27cfd23_1
  mkl_fft            pkgs/main/linux-64::mkl_fft-1.3.0-py38h42c9631_2
  mkl_random         pkgs/main/linux-64::mkl_random-1.2.1-py38ha9443f7_2
  ncurses            pkgs/main/linux-64::ncurses-6.2-he6710b0_1
  ninja              pkgs/main/linux-64::ninja-1.10.2-hff7bd54_1
  numpy              pkgs/main/linux-64::numpy-1.20.1-py38h93e21f0_0
  numpy-base         pkgs/main/linux-64::numpy-base-1.20.1-py38h7d8b39e_0
  openssl            pkgs/main/linux-64::openssl-1.1.1k-h27cfd23_0
  pip                pkgs/main/linux-64::pip-21.0.1-py38h06a4308_0
  python             pkgs/main/linux-64::python-3.8.8-hdb3f193_5
  pytorch            pytorch-nightly/linux-64::pytorch-1.9.0.dev20210428-py3.8_cuda11.1_cudnn8.0.5_0
  readline           pkgs/main/linux-64::readline-8.1-h27cfd23_0
  setuptools         pkgs/main/linux-64::setuptools-52.0.0-py38h06a4308_0
  six                pkgs/main/linux-64::six-1.15.0-py38h06a4308_0
  sqlite             pkgs/main/linux-64::sqlite-3.35.4-hdfb4753_0
  tk                 pkgs/main/linux-64::tk-8.6.10-hbc83047_0
  typing_extensions  pkgs/main/noarch::typing_extensions-3.7.4.3-pyha847dfd_0
  wheel              pkgs/main/noarch::wheel-0.36.2-pyhd3eb1b0_0
  xz                 pkgs/main/linux-64::xz-5.2.5-h7b6447c_0
  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3


Preparing transaction: done
Verifying transaction: done
Executing transaction: \ By downloading and using the CUDA Toolkit conda packages, you accept the terms and conditions of the CUDA End User License Agreement (EULA): https://docs.nvidia.com/cuda/eula/index.html

done
#
# To activate this environment, use
#
#     $ conda activate /tmp/tmp.A1nJh47HsJ
#
# To deactivate an active environment, use
#
#     $ conda deactivate


/tmp/tmp.A1nJh47HsJ 15s
/tmp/tmp.A1nJh47HsJ ❯ conda activate $(pwd)

/tmp/tmp.A1nJh47HsJ
/tmp/tmp.A1nJh47HsJ ❯ python -c 'import torch'
```

</details>

**Comment by rafi-cohen** [comment]
> Regarding the original failure this seems to be related to conda-forge, when I did a `s/conda-forge/nvidia/` to pull `cudatoolkit=11.1` from `-c nvidia` it appears to work:

Replacing `conda-forge` with `nvidia` indeed seems to work. However, I'm not sure the problem is the `cudatoolkit` package. The issue still occurs if `cudatoolkit` is grabbed from `nvidia`, but other packages from `conda-forge`.

<details>
  <summary>Click for logs</summary>
  

```
$ conda create -y -n nightly python=3.8 pytorch cudatoolkit=11.1 -c pytorch-nightly -c nvidia -c conda-forge && conda activate nightly && python -c "import torch"
Collecting package metadata (current_repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.9.2
  latest version: 4.10.1

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: path/to/envs/nightly

  added / updated specs:
    - cudatoolkit=11.1
    - python=3.8
    - pytorch


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_llvm
  blas               pkgs/main/linux-64::blas-1.0-mkl
  ca-certificates    conda-forge/linux-64::ca-certificates-2020.12.5-ha878542_0
  certifi            conda-forge/linux-64::certifi-2020.12.5-py38h578d9bd_1
  cudatoolkit        nvidia/linux-64::cudatoolkit-11.1.74-h6bb024c_0
  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.35.1-hea4e1c9_2
  libffi             conda-forge/linux-64::libffi-3.3-h58526e2_2
  libgcc-ng          conda-forge/linux-64::libgcc-ng-9.3.0-h2828fa1_19
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-9.3.0-h6de172a_19
  libuv              conda-forge/linux-64::libuv-1.41.0-h7f98852_0
  llvm-openmp        conda-forge/linux-64::llvm-openmp-11.1.0-h4bd325d_1
  mkl                conda-forge/linux-64::mkl-2020.4-h726a3e6_304
  mkl-service        conda-forge/linux-64::mkl-service-2.3.0-py38h1e0a361_2
  mkl_fft            conda-forge/linux-64::mkl_fft-1.3.0-py38h5c078b8_1
  mkl_random         conda-forge/linux-64::mkl_random-1.2.0-py38hc5bc63f_1
  ncurses            conda-forge/linux-64::ncurses-6.2-h58526e2_4
  ninja              conda-forge/linux-64::ninja-1.10.2-h4bd325d_0
  numpy              pkgs/main/linux-64::numpy-1.19.2-py38h54aff64_0
  numpy-base         pkgs/main/linux-64::numpy-base-1.19.2-py38hfa32c7d_0
  openssl            conda-forge/linux-64::openssl-1.1.1k-h7f98852_0
  pip                conda-forge/noarch::pip-21.1-pyhd8ed1ab_0
  python             conda-forge/linux-64::python-3.8.8-hffdb5ce_0_cpython
  python_abi         conda-forge/linux-64::python_abi-3.8-1_cp38
  pytorch            pytorch-nightly/linux-64::pytorch-1.9.0.dev20210428-py3.8_cuda11.1_cudnn8.0.5_0
  readline           conda-forge/linux-64::readline-8.1-h46c0cb4_0
  setuptools         conda-forge/linux-64::setuptools-49.6.0-py38h578d9bd_3
  six                conda-forge/noarch::six-1.15.0-pyh9f0ad1d_0
  sqlite             conda-forge/linux-64::sqlite-3.35.5-h74cdb3f_0
  tk                 conda-forge/linux-64::tk-8.6.10-h21135ba_1
  typing_extensions  conda-forge/noarch::typing_extensions-3.7.4.3-py_0
  wheel              conda-forge/noarch::wheel-0.36.2-pyhd3deb0d_0
  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1
  zlib               conda-forge/linux-64::zlib-1.2.11-h516909a_1010


Preparing transaction: done
Verifying transaction: done
Executing transaction: \ By downloading and using the CUDA Toolkit conda packages, you accept the terms and conditions of the CUDA End User License Agreement (EULA): https://docs.nvidia.com/cuda/eula/index.html

done
#
# To activate this environment, use
#
#     $ conda activate nightly
#
# To deactivate an active environment, use
#
#     $ conda deactivate

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "path/to/envs/nightly/lib/python3.8/site-packages/torch/__init__.py", line 196, in <module>
    _load_global_deps()
  File "path/to/envs/nightly/lib/python3.8/site-packages/torch/__init__.py", line 149, in _load_global_deps
    ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
  File "path/to/envs/nightly/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libmkl_intel_lp64.so.1: cannot open shared object file: No such file or directory
```

</details>

**Comment by asnota** [comment]
I'm not using conda in Colab environment, however I have the same issue - it appeared on Friday 23 of April and is not resolved yet. In my case, I need to use XLA and to install it, a nighty version of pytorch is used: 
`!curl https://raw.githubusercontent.com/pytorch/xla/master/contrib/scripts/env-setup.py -o pytorch-xla-env-setup.py
!python pytorch-xla-env-setup.py --version nightly --apt-packages libomp5 libopenblas-dev`

with the installation log:
`Installing collected packages: torch
Successfully installed torch-1.9.0a0+git77721ee
Processing ./torch_xla-nightly-cp37-cp37m-linux_x86_64.whl
Installing collected packages: torch-xla
Successfully installed torch-xla-1.9+759e765
Processing ./torchvision-nightly-cp37-cp37m-linux_x86_64.whl
Requirement already satisfied: pillow>=5.3.0 in /usr/local/lib/python3.7/dist-packages (from torchvision==nightly) (7.1.2)
Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from torchvision==nightly) (1.19.5)
Requirement already satisfied: torch in /usr/local/lib/python3.7/dist-packages (from torchvision==nightly) (1.9.0a0+git77721ee)
Requirement already satisfied: typing-extensions in /usr/local/lib/python3.7/dist-packages (from torch->torchvision==nightly) (3.7.4.3)
Installing collected packages: torchvision
Successfully installed torchvision-0.10.0a0+54d88f7
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libopenblas-dev is already the newest version (0.2.20+ds-4).
The following package was automatically installed and is no longer required:
  libnvidia-common-460
Use 'apt autoremove' to remove it.
The following NEW packages will be installed:
  libomp5
0 upgraded, 1 newly installed, 0 to remove and 34 not upgraded.
Need to get 234 kB of archives.
After this operation, 774 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libomp5 amd64 5.0.1-1 [234 kB]
Fetched 234 kB in 1s (401 kB/s)
Selecting previously unselected package libomp5:amd64.
(Reading database ... 161547 files and directories currently installed.)
Preparing to unpack .../libomp5_5.0.1-1_amd64.deb ...
Unpacking libomp5:amd64 (5.0.1-1) ...
Setting up libomp5:amd64 (5.0.1-1) ...
Processing triggers for libc-bin (2.27-3ubuntu1.2) ...
/sbin/ldconfig.real: /usr/local/lib/python3.7/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link`

the attempt to import xla then causes the following error:
`OSError                                   Traceback (most recent call last)
<ipython-input-7-7fe098b5b807> in <module>()
----> 1 import torch_xla

3 frames
/usr/lib/python3.7/ctypes/__init__.py in __init__(self, name, mode, handle, use_errno, use_last_error)
    362 
    363         if handle is None:
--> 364             self._handle = _dlopen(self._name, mode)
    365         else:
    366             self._handle = handle

OSError: libmkl_intel_lp64.so.1: cannot open shared object file: No such file or directory`

**Comment by malfet** [comment]
Ok, this happens because pytorch package is getting built against `mkl-2021.2` from conda-forget, which contains those `.so.1` links, compared to `mkl-2020.2` available from conda.

**Comment by malfet** [comment]
This should fix it: https://github.com/pytorch/builder/pull/750

**Comment by malfet** [comment]
Was broken  1.9.0.dev20210429-py3.7_cpu_0, fixed in 1.9.0.dev20210503-py3.7_cpu_0, closing.
